### PR TITLE
Mid: extra: Change pcmk_snmp_helper.sh

### PR DESCRIPTION
**Pull request by HideoYamauchi**
Hi All,

I send the script for the SNMP trap
I branch off the transmission in a parameter.

We want to put this script as the script which is available to users immediately.

We want to include this script in Pacemaker1.1.15 with order of traps control.

The script of Alert is different from the OCF resource, but thinks that Pacemaker should add it to the head of the environment variable handed in instance_attributes what it is. ex) SCRIPT_xxxx.

(This patch reconstituted next pullrequest.- https://github.com/ClusterLabs/pacemaker/pull/1006)

Best Regards,
Hideo Yamauchi.

### CODE REVIEW

**Comment by HideoYamauchi** [comment]
Hi All,

Should snmptrap be able to appoint the v option in a parameter, too?

Best Regards,
Hideo Yamauchi.

**Comment by wenningerk** [comment]
Had done some necro-posting in the #1006 pull-request - so copy here...
(I have no strong opinion regarding the trap_version. But with v1 the cmdline would have
to look differently and to exploit the benefits of v3 we would probably need user and password) :

Hi Hideo,

the attributes are still in the alerts-block instead of the alert-block ;-)

I had got that you didn't see the hrSystemDate OID timestamp as a solution to your problem.
But having seen that loosing the order of traps might be a general issue I still see it as the only easy
approach when traps are coming from different nodes and/or if you want to avoid the disadvantages
of the queueing approach. This might for some reasons not apply for
your problem but to give a generic solution it is still a way to go and it is a good example
to show the possibilities of the interface enhancements.
So I think we should have the timestamp-OID in. Can be configurable via a switch as the
traps...

It is quite common in the cib to express boolean decisions via true/false - so this might
be nicer here as well - just for consistency.

Your implementation btw. is kind of talkative as it sends traps for all monitor-actions.
This might be interesting for certain cases but I guess for the default case the original
approach just to trap failed monitors would be the better solution. Thus a switch like
trap_all_monitors would be handy here as well.

Regards,
Klaus

**Comment by HideoYamauchi** [comment]
Hi Klaus,

> Had done some necro-posting in the #1006 pull-request - so copy here...
> (I have no strong opinion regarding the trap_version. But with v1 the cmdline would have
> to look differently and to exploit the benefits of v3 we would probably need user and password) :

I investigate some snmptrap.
If it is necessary, I add user and passwd.

> the attributes are still in the alerts-block instead of the alert-block ;-)

Does it mean that the meaning that you say may be the next?
```
# ================================
# <configuration>
#   <alerts>
#     <alert id="snmp_alert" path="/path/to/pcmk_snmp_notify.sh">
#       <instance_attributes id="insta_9">
#         <nvpair id="trap_nodes" name="trap_node" value="no"/>
#         <nvpair id="trap_fencing" name="trap_fencing" value="no"/>
#       </instance_attributes>
#       <recipient id="snmp_destination" value="192.168.1.2"/>
#     </alert>
#   </alerts>
# </configuration>
# ================================
 or 
# ================================
# <configuration>
#   <alerts>
#     <alert id="snmp_alert" path="/path/to/pcmk_snmp_notify.sh">
#       <recipient id="snmp_destination" value="192.168.1.2">
#        <instance_attributes id="insta_9">
#         <nvpair id="trap_nodes" name="trap_node" value="no"/>
#         <nvpair id="trap_fencing" name="trap_fencing" value="no"/>
#        </instance_attributes>
#       </recipient>
#     </alert>
#   </alerts>
# </configuration>
# ================================

```
I misunderstood it a little and seemed to write a sample.
I revise a sample.

>I had got that you didn't see the hrSystemDate OID timestamp as a solution to your problem.
>But having seen that loosing the order of traps might be a general issue I still see it as the only easy
>approach when traps are coming from different nodes and/or if you want to avoid the disadvantages
>of the queueing approach. This might for some reasons not apply for
>your problem but to give a generic solution it is still a way to go and it is a good example
>to show the possibilities of the interface enhancements.
>So I think we should have the timestamp-OID in. Can be configurable via a switch as the
>traps...

I was not able to understand this opinion well.
Is it to say that it is necessary to set "hrSystemDate OID timestamp" from a script?
I do not have a lot on specifications of the SNMP.
Or do I misunderstand it really?

But...I make a patch of "queue" of the trap of crmd now.
(This patch resembles the patch which I contributed in old days.)

> It is quite common in the cib to express boolean decisions via true/false - so this might
> be nicer here as well - just for consistency.

The parameter revises it for handling of "true/false".

> Your implementation btw. is kind of talkative as it sends traps for all monitor-actions.
> This might be interesting for certain cases but I guess for the default case the original
> approach just to trap failed monitors would be the better solution. Thus a switch like
> trap_all_monitors would be handy here as well.

I would have a talkative script to omit trouble to customize the script of the user if possible.
Leaving an original script, there is no problem if I have you add this script.

Best Regards,
HIdeo Yamauchi.

**Comment by wenningerk** [comment]
Either of your xml-examples should conform with the schema and in the case
that there is just one recipient section there is no difference in which environment
variables the script is being called with.
As it is the more general I would go with the 2nd example.
If you have 2 recipients you could of course by having the attributes in the recipient-section
achieve that e.g. the node-alerts go to one recipient while the resource-alerts go to another.

The snmptrap-tool takes a list of "{OID} {TYPE} {VALUE}"-triples.
What I meant was to add hrSystemDate to that list.  

Haven't tried it but I would try to go with something like this:

set meta-attribute timestamp-format="%Y-%m-%d,%H:%M:%S.%01N"
and add

  HOST-RESOURCES-MIB::hrSystemDate s "$CRM_alert_timestamp"

to the snmptrap-tool parameters.

Regarding your last comment I don't understand it.
What I meant was to add an additional instance-attribute to be able to switch on sending 
of traps for resource-monitors that had not failed. Per default it would just send traps for
monitors that had gone wrong - as it does as of today.

Regards,
Klaus

**Comment by HideoYamauchi** [comment]
Hi Klaus,

The setting of the sample of both scripts include it.
```
# ================================
# <configuration>
#   <alerts>
#     <alert id="snmp_alert" path="/path/to/pcmk_snmp_notify.sh">
#       <instance_attributes id="insta_9">
#         <nvpair id="trap_nodes" name="trap_node" value="no"/>
#         <nvpair id="trap_fencing" name="trap_fencing" value="no"/>
#       </instance_attributes>
#       <recipient id="snmp_destination" value="192.168.1.2"/>
#     </alert>
#   </alerts>
# </configuration>
# ================================
# ================================
# <configuration>
#   <alerts>
#     <alert id="snmp_alert" path="/path/to/pcmk_snmp_notify.sh">
#       <recipient id="snmp_destination" value="192.168.1.2">
#        <instance_attributes id="insta_9">
#         <nvpair id="trap_nodes" name="trap_node" value="no"/>
#         <nvpair id="trap_fencing" name="trap_fencing" value="no"/>
#        </instance_attributes>
#       </recipient>
#     </alert>
#   </alerts>
# </configuration>
# ================================
```

> Regarding your last comment I don't understand it.
> What I meant was to add an additional instance-attribute to be able to switch on sending 
> of traps for resource-monitors that had not failed. Per default it would just send traps for
> monitors that had gone wrong - as it does as of today.

Okay!
I add the parameter that the trap of monitor not to fail in does not transmit.
(example:instance_attributes name is only_monitor_failed.)

```
<nvpair id="only_monitor_failed" name="only_monitor_failed" value="true"/>
```

The judgment of the former script serves as a reference.

```
(snip)
if [[ ${CRM_alert_rc} != 0 && ${CRM_alert_task} == "monitor" ]] || [[ ${CRM_alert_task} != "monitor" ]] ; then
(snip)
```


I will confirm your other opinions again tomorrow.

Best Regards,
Hideo Yamauchi.

File: extra/alerts/pcmk_snmp_helper.sh [hunk 1]
```diff
@@ -1,56 +1,94 @@
...
-    /usr/bin/snmptrap -v 2c -c public ${CRM_alert_recipient} "" PACEMAKER-MIB::pacemakerNotificationTrap \
-	PACEMAKER-MIB::pacemakerNotificationNode s "${CRM_alert_node}" \
-	PACEMAKER-MIB::pacemakerNotificationResource s "${CRM_alert_rsc}" \
-	PACEMAKER-MIB::pacemakerNotificationOperation s "${CRM_alert_task}" \
-	PACEMAKER-MIB::pacemakerNotificationDescription s "${CRM_alert_desc}" \
-	PACEMAKER-MIB::pacemakerNotificationStatus i "${CRM_alert_status}" \
-	PACEMAKER-MIB::pacemakerNotificationReturnCode i ${CRM_alert_rc} \
-	PACEMAKER-MIB::pacemakerNotificationTargetReturnCode i ${CRM_alert_target_rc} && exit 0 || exit 1
+if [ -z $CRM_alert_version ]; then
+    echo "Pacemaker version 1.1.15 is required"
```
Comment (kgaillot) [inline extra/alerts/pcmk_snmp_helper.sh]: "1.1.15 or later"

Comment (HideoYamauchi) [inline extra/alerts/pcmk_snmp_helper.sh]: Okay!

**Comment by HideoYamauchi** [comment]
Hi Klaus,

> The snmptrap-tool takes a list of "{OID} {TYPE} {VALUE}"-triples.
> What I meant was to add hrSystemDate to that list.
>
> Haven't tried it but I would try to go with something like this:
> 
> set meta-attribute timestamp-format="%Y-%m-%d,%H:%M:%S.%01N"
> and add
> 
> HOST-RESOURCES-MIB::hrSystemDate s "$CRM_alert_timestamp"
> 
> to the snmptrap-tool parameters.

It was understood that I set "$CRM_alert_timestamp" in "HOST-RESOURCES-MIB::hrSystemDate" in snmptrap.

However, there is no "HOST-RESOURCES-MIB::hrSystemDate" in PACEMAKER-MIB.

Does it mean that I change PACEMAKER-MIB?
Or do you mean other methods?

Best Regards,
Hideo Yamauchi.

**Comment by wenningerk** [comment]
Hi Hideo!

hrSystemDate is part of HOST-RESOURCES-MIB and therefore and as being something
generic wouldn't belong into PACEMAKER-MIB.
I checked on a RHEL-7.2 and there it comes via package net-snmp-libs and is located
/usr/share/snmp/mibs/HOST-RESOURCES-MIB.txt.
So I guess it is quite standard to have it on a system.
Still we can make it optionally switchable via instance-attribute and/or hardcode it in
the script - 1.3.6.1.2.1.25.1.2.

Regards,
Klaus

**Comment by HideoYamauchi** [comment]
Hi Klaus,

I understood somehow.
Please point it out if wrong.


I think that it is to say to use HOST-RESOURCES-MIB.
A script sends the value as if Pacemaker being able to generate the value for HOST-RESOURCES-MIB.

For example, is it the following method?

```
if [ use HOST-RESOURCES-MIB] then;
   ${trap_binary} -v 2c -c ${trap_community} ${CRM_alert_recipient} ""  HOST-RESOURCES-MIB.....
(snip)
fi;

if [ use PACEMAKER-MIB ] then;
case $CRM_alert_kind in
    node)
        if [ ${trap_node} = "yes" ]; then
            ${trap_binary} -v 2c -c ${trap_community} ${CRM_alert_recipient} "" PACEMAKER-MIB::pacemakerNotificationTrap \
                PACEMAKER-MIB::pacemakerNotificationNode s "${CRM_alert_node}" \                PACEMAKER-MIB::pacemakerNotificationDescription s "${CRM_alert_desc}"        fi
        ;;
    fencing)
(snip)

```

I made the patch which I added queue function to.
I test it from now on.
I am going to send the patch next Tuesday.


Best Regards,
Hideo Yamauchi.

**Comment by wenningerk** [comment]
It should be possible to just send everything in one trap otherwise we would 
not really gain any benefit.

**Comment by HideoYamauchi** [comment]
Hi Klaus,

I do not know whether snmptrap command can transmit HOST-RESOURCES-MIB and PACEMAKER-MIB at the same time.
I check it a little.

Many thanks!

**Comment by wenningerk** [comment]
Hi Hideo!

It shouldn't care what OIDs it is sending. Actually it should just convert the string to the numeric 
representation and have no real knowledge what stands behind it. You can even give the
numeric representation directly.

Regards,
Klaus

**Comment by HideoYamauchi** [comment]
Hi Klaus,

All right!

I concentrate power on the making of the patch of queue.
I intend to send the patch which the opinion to this script reflected next week again.

Best Regards,
Hideo Yamauchi.

**Comment by HideoYamauchi** [comment]
Hi Klaus,
Hi Ken,

I send new pullrequest of the revision.
I close this old pullrequest.

Many thanks,
Hideo Yamauchi.

**Pull request closed without merging**